opacus/grad_sample/grad_sample_module.py

```diff
@@ -425,6 +425,18 @@ def rearrange_grad_samples(
 
         return activations, backprops
 
+    def state_dict(self, *args, **kwargs):
```

Comment:
Let's reopen this conversation: do we want to reroute *all* unknown (i.e. not explicitly declared in GradSampleModule) attributes to the underlying `_module`?
Right now we only do this for named parameters (see `__getattr__`) .

The reason I'm raising this here is because with new `state_dict` and `load_state_dict` we create a discrepancy between the output of `state_dict` and the output of `named_parameters()` method - which can produce unexpected results.
I think at least we should also override all `named_*` method here too, and at most - think about being more radical and just updating `__getattr__`.

Full context: I was planning to override `named_*` method for other reasons, just didn't get to publishing a PR yet

Reply:
>  do we want to reroute all unknown (i.e. not explicitly declared in GradSampleModule) attributes to the underlying _module?

Good point! Perhaps not! Also, I realise now that we should also store GS specific attributes such as `loss_reduction` and `batch_first` in the `GS.state_dict()`

>  discrepancy between the output of state_dict and the output of named_parameters()

The only difference right now is the extra `_module` prefix, correct?

Reply:
> Good point! Perhaps not!

Or perhaps yes...

> I realise now that we should also store GS specific attributes such as loss_reduction and batch_first in the GS.state_dict()

Should we though? Are those fundamentally part of the model state or model architecture?

> The only difference right now is the extra _module prefix, correct?

Affirmative

Reply:
> Should we though? Are those fundamentally part of the model state or model architecture?

They are not, but are essential to instantiate a GS module. I don't see a downside if we do. Do you feel otherwise?